### One Hot Encoding - variables with many categories
one hot encoding is used for converting categorical data into numerical data

In [40]:
import pandas as pd
import numpy as np

# loading mercedes dataset
data = pd.read_csv("mercedes.csv", usecols = ['X1', 'X2', 'X3', 'X4', 'X5', 'X6'])
data.head()


X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [41]:
# lets check how many unique values values are present in each column

for col in data.columns:
    print(col, ':', len(data[col].unique()), 'labels')


X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels


In [42]:
# let's examine how many columns we will obtain after one hot encoding these variables

pd.get_dummies(data, drop_first = True).shape

(4209, 117)

We can see that from just 6 initial categorical variables, we end up with 117 new variables.

In [43]:
# let's find the top 10 most frequent categories for the variables x2

data.X2.value_counts().sort_values(ascending = False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: X2, dtype: int64

In [44]:
# lets make a list with the most frequent categoris for the variabls

top_10 = [x for x in data.X2.value_counts().sort_values(ascending = False).head(10).index]
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [45]:
# and now we make the 10 binary variables

for label in top_10:
    data[label] = np.where(data['X2'] == label, 1, 0)
data[['X2'] + top_10].head(20)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0

In [47]:
import pandas as pd
import numpy as np

# Read the file again
data = pd.read_csv('mercedes.csv', usecols=['X1', 'X2', 'X3', 'X4', 'X5', 'X6'])

# Determine the top 10 most frequent labels in the 'X2' column
top_10 = data['X2'].value_counts().head(10).index

# Function to create the dummy variables for the most frequent labels
def one_hot_top_x(df, variable, top_x_labels):
    for label in top_x_labels:
        df[variable + '_' + label] = np.where(df[variable] == label, 1, 0)

# Encode 'X2' into the 10 most frequent categories
one_hot_top_x(data, 'X2', top_10)
data.head()

    
 

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

In [38]:
# find the 10 most frequent categoris for x1
top_10 = [x for x in data.X1.value_counts().sort_values(ascending = False).head(10).index]

# now create the 10 most frequent dummy variables for x1
one_hot_top_x(data, 'X1', top_10)
data.head()

TypeError: 'DataFrame' object is not callable